Allow empty or 'none' as audio.mixer value

ZenithDK opened the pull request:
To disable mixing altogether, you can now let the configuration value
audio/mixer be empty or set it to 'none'.

Fixes #936

adamcik commented:
Looks good, only change I want to suggest is that we not make the config value optional. Better to have it explicitly be set to `none` IMO.

Also did you do any quick hands on testing with MPC or some other client with the mixer set to none or are we just assuming the unit tests cover not having a mixer set well enough?

ZenithDK commented:
I tested with a web client and while I can see the requests going in via JSON, nothing changed in the model and the volume stays the same.
I will change to making it explicit and not optional.
I would like to implement unit tests for ``none`` but I'm unsure of how I would go about that.

adamcik commented:
For core and the WS case you can add a new test class to https://github.com/mopidy/mopidy/blob/develop/tests/core/test_mixer.py

For the MPD case the test cases you probably want to create a new no-mixer-textcasess looking at the following tests for inspiration https://github.com/mopidy/mopidy/blob/develop/tests/mpd/protocol/test_playback.py#L81

And https://github.com/mopidy/mopidy/blob/develop/tests/mpd/protocol/__init__.py#L35 probably needs to be updated to have a way to opt out of having a mixer by setting some attribute or something.

ZenithDK commented:
Fixed

adamcik commented on tests/core/test_mixer.py:
```diff
@@ -33,3 +33,23 @@ def test_set_mute(self):
+
+class CoreMixerNone(unittest.TestCase):
+    def setUp(self):  # noqa: N802
+        self.mixer = None
```
Might as well just inline the `None` in this case.

adamcik commented on tests/mpd/protocol/__init__.py:
```diff
@@ -80,3 +80,19 @@ def assertNotInResponse(self, value):  # noqa: N802
         self.assertEqual(value, self.connection.response[0])
+
+
+class BaseTestCaseMixerNone(BaseTestCase):
```
Don't bother with using the `BaseTest` pattern here as we only have the one test case using this setup logic.

What I was thinking of was more something like the following snippet, mainly because it saves you having to re-implement the rest of the setup method.

```
class BaseTestCase(unittest.TestCase):
    enable_mixer = True

    def setUp(self):  # noqa: N802
        if self.enable_mixer:
            self.mixer = dummy_mixer.create_proxy()
        else:
            self.mixer = None
```

adamcik commented on tests/mpd/protocol/test_playback.py:
```diff
@@ -463,3 +463,20 @@ def test_stop(self):
+    def test_setvol_min(self):
+        self.send_request('setvol "0"')
+        self.assertEqual(None, self.core.mixer.get_volume().get())
+        self.assertInResponse('OK')
```
```
setvol 0
ACK [52@0] {setvol} problems setting volume
```

adamcik commented:
Did a quick test with an MPD without a mixer and go the following:

```
setvol 0
ACK [52@0] {setvol} problems setting volume
status
volume: -1
repeat: 0
random: 0
single: 0
consume: 0
playlist: 2
playlistlength: 14
mixrampdb: 0.000000
state: stop
song: 0
songid: 1
nextsong: 1
nextsongid: 2
```

jodal commented on mopidy/commands.py:
```diff
@@ -306,13 +309,18 @@ def get_mixer_class(self, config, mixer_classes):
                 config['audio']['mixer'],
-                ', '.join([m.name for m in mixer_classes]))
+                ', '.join([m.name for m in mixer_classes]) + ', none' or
+                'none')
```
Maybe a bit clever, but I'll just add the trick to my book and it will become familiar ;-)

ZenithDK replied:
It was that or a much larger rewrite :-) Hackish, but working ;)

jodal commented:
Looks good to me. Nothing to add to @adamcik's comments.

ZenithDK commented:
@adamcik: I need to change the handling of volume with the ``none`` mixer to return an error in the ACK?
Also, not sure how set ``enable_mixer`` to False in your code before the ``setUp()`` function is called?

adamcik commented:
Something like:

```
class FooTest(Base...):
    enable_mixer = False

    def test....(...):
        ....
```

The mpd volume handling needs to handle convert a `None` volume to `-1` to cover the easy one first. For the other setting the volume we probably need to update the core APIs to return `True/False` so we can return an MPD error when appropriate.

I also just remembered that we have mute handling hackend into "outputs", we might want to test that as well.

Sorry for turning this into a yak shaving exercise :(

ZenithDK commented:
Updated the API to return True/False and implemented the MPD error and test cases.
Also implemented support for ``toggleoutput`` and added tests for that. I will probably look at increasing test coverage some more, py.test seems pretty good, I like how it displays the results.

jodal commented on mopidy/commands.py:
```diff
@@ -306,13 +309,18 @@ def get_mixer_class(self, config, mixer_classes):
             ', '.join(m.__name__ for m in mixer_classes) or 'none')
 
+        if config['audio']['mixer'] == 'none':
+            logger.debug("Mixer disabled")
```
Prefer single quotes

jodal commented on mopidy/core/mixer.py:
```diff
@@ -21,7 +22,9 @@ def get_volume(self):
         """
-        if self._mixer is not None:
+        if self._mixer is None:
+            return -1
```
The behavior here no longer matches the docstring. Why not continue to return `None` if the volume is unknown?

jodal commented on mopidy/core/mixer.py:
```diff
@@ -31,8 +34,10 @@ def set_volume(self, volume):
 
         The volume scale is linear.
```
This docstring must be updated with the return values.

ZenithDK replied:
Fixed all of the docstring updates in latest update.

jodal commented on mopidy/core/mixer.py:
```diff
@@ -48,5 +55,10 @@ def set_mute(self, mute):
 
         :class:`True` to mute, :class:`False` to unmute.
         """
```
This docstring must be updated with the return values.

jodal commented on mopidy/mpd/protocol/__init__.py:
```diff
@@ -23,7 +23,7 @@
 LINE_TERMINATOR = '\n'
 
 #: The MPD protocol version is 0.17.0.
```
Comment doesn't match next line.

ZenithDK replied:
Reverted this, as I am not sure if increasing it has any unintended side effects with some clients, and it was for testing why "changed: outputs" suddenly stopped working with the none mixer.

jodal commented on mopidy/mpd/protocol/playback.py:
```diff
@@ -397,7 +397,8 @@ def setvol(context, volume):
     """
     # NOTE: we use INT as clients can pass in +N etc.
-    context.core.mixer.set_volume(min(max(0, volume), 100))
+    if context.core.mixer.set_volume(min(max(0, volume), 100)).get() is False:
```
What about this instead?

``` python
success = context.core.mixer.set_volume(min(max(0, volume), 100)).get()
if not success:
    # ...
```

jodal replied:
Maybe even extract `volume = min(max(0, volume), 100)` as a variable of its own.

jodal commented on tests/core/test_mixer.py:
```diff
@@ -33,3 +33,21 @@ def test_set_mute(self):
+        self.core = core.Core(None, backends=[])
+
+    def test_get_volume(self):
+        self.assertEqual(self.core.mixer.get_volume(), -1)
```
I think this should return `None`?

jodal commented on tests/core/test_mixer.py:
```diff
@@ -33,3 +33,21 @@ def test_set_mute(self):
+
+    def test_set_mute(self):
+        self.core.mixer.set_mute(True)
+        self.assertEqual(self.core.mixer.set_mute(True), True)
```
The test name should tell more about what behavior you expect here. Currently, it is only defined in the implementation, not in tests or docstrings.

jodal commented on tests/mpd/protocol/test_playback.py:
```diff
@@ -150,6 +150,14 @@ def test_replay_gain_status_default(self):
 
+    def test_mixrampdb(self):
+        self.send_request('mixrampdb "10"')
+        self.assertInResponse('OK')
```
Shouldn't we return a not implemented error here?

jodal commented on tests/mpd/protocol/test_playback.py:
```diff
@@ -150,6 +150,14 @@ def test_replay_gain_status_default(self):
+
+    def test_mixrampdelay(self):
+        self.send_request('mixrampdelay "10"')
+        self.assertInResponse('OK')
```
And here?

ZenithDK commented:
Updated with more tests and reworking here and there.
I updated the DummyMixer so that it also calls the MixerListener events now so we can test for this and made it return the proper return codes since I updated that per @adamcik's comments.

adamcik commented on mopidy/mpd/protocol/audio_output.py:
```diff
@@ -43,7 +42,11 @@ def toggleoutput(context, outputid):
         Turns an output on or off, depending on the current state.
     """
-    pass
+    if outputid == 0:
```
Cool that you are fixing this, but next time I would make it a separate commit within the PR as this is a distinct change from the main thing you are changing in the PR.

adamcik commented on mopidy/mpd/protocol/playback.py:
```diff
@@ -46,11 +45,10 @@ def mixrampdb(context, decibels):
     tags crossfading will be used. See http://sourceforge.net/projects/mixramp
     """
-    pass
+    raise exceptions.MpdNotImplemented  # TODO
```
Same goes for these, that is I would have preferred that these be a separate commit with just "Make mixramp commands raise not implemented" instead of squashing everything into a single commit. And just to be clear I don't expect you to fix this for this PR/commit, just keep it in mind next time around :-)

adamcik commented on tests/core/test_listener.py:
```diff
@@ -57,3 +57,6 @@ def test_listener_has_default_impl_for_mute_changed(self):
     def test_listener_has_default_impl_for_seeked(self):
         self.listener.seeked(0)
+
+    def test_listener_has_default_impl_for_current_metadata_changed(self):
```
Heh, this will likely break with my other changes I have underway, guess we have to see who of us can get the change in first :P

adamcik commented on docs/changelog.rst:
```diff
@@ -113,6 +116,9 @@ v0.20.0 (UNRELEASED)
   :meth:`mopidy.core.LibraryController.get_distinct`. (Fixes: :issue:`913`)
 
+- Add support for ``toggleoutput`` command. ``mixrampdb`` and ``mixrampdelay``
+  have also been added but have not been implemented yet.
```
This is a bit of a nitpick, but I wouldn't write "not been implemented yet" as it implies we might add support for it. At this time I have no intention of adding Gnonlin as a dependency just so people can get crossfaded songs.

ZenithDK replied:
"The ``mixrampdb`` and ``mixrampdelay`` commands are now supported but throw a ``NotImplemented`` exception"?

adamcik commented on mopidy/core/mixer.py:
```diff
@@ -40,13 +44,22 @@ def get_mute(self):
-        if self._mixer is not None:
-            self._mixer.set_mute(bool(mute))
+        if self._mixer is None:
+            self._mute = mute
```
Why do we track the mute state when there is no mixer?

ZenithDK replied:
Because you can mute the output - and the state is needed when ``get_mute`` is called. The ``mute_changed`` event could get the value directly from the parameters, though.

ZenithDK replied:
I got "output" (from MPD) mixed up with "mute" in mopidy - so I'm removing this and just returning ``False`` always. The mpd frontend will translate this into an exception.

ZenithDK commented on tests/mpd/protocol/test_idle.py:
```diff
@@ -102,6 +108,22 @@ def test_idle_player_then_event_player(self):
+        self.assertNoEvents()
+        self.assertOnceInResponse('changed: player')
+        self.assertOnceInResponse('OK')
+
```
A copy-paste error here, should be "output" of course. Fixed that in latest commit.

ZenithDK commented on tests/mpd/protocol/test_idle.py:
```diff
@@ -206,3 +228,36 @@ def test_player_then_playlist_then_idle_playlist(self):
+        self.assertNoEvents()
+        self.assertNoSubscriptions()
+        self.assertOnceInResponse('changed: output')
+        self.assertOnceInResponse('OK')
```
As we don't want to be able to mute/unmute (or enable/disable in MPD speak) - I have removed this as there should not be any ``changed: output`` events sent any longer.

ZenithDK commented:
Updated changelog and removed tracking of the mute status and cleaned up the testcases some more, and fixed a few bugs in the testcases.
